velox/expression/ComplexViewTypes.h

```diff
@@ -64,65 +60,81 @@ class IndexBasedIterator
   Iterator& operator++() {
-    element_.incrementIndex();
+    index_++;
     return *this;
   }
 
  protected:
-  T element_;
+  vector_size_t index_;
 };
 
-// Implements an iterator for T::element_t that moves by calling
-// incrementIndex() until it points to the next not-null element. T is expected
-// to have index(), incrementIndex(), value(), and has_value().
+// Implements an iterator for values stored in the reader T
+// that skip nulls and provides direct access to the value.
 template <typename T>
-class SkipNullsIterator
-    : public std::
-          iterator<std::input_iterator_tag, typename T::element_t, size_t> {
-  using Iterator = SkipNullsIterator<T>;
-  using value_type = typename T::element_t;
+class SkipNullsIterator;
+
+template <typename T>
+class SkipNullsIterator {
+  // Here value_type is either a primitive, StringView, ArrayView, MapView, or
+  // RowView.
+  using value_type = typename T::exec_in_t;
+  using reference = value_type;
+  using pointer = value_type;
+  using difference_type = size_t;
+  using iterator_category = std::input_iterator_tag;
 
  public:
-  SkipNullsIterator<T>(const T& element, vector_size_t lasIndex)
-      : element_(element), endIndex_(lasIndex) {}
+  using Iterator = SkipNullsIterator<T>;
+  SkipNullsIterator<T>(
+      const T* reader,
+      vector_size_t index,
+      vector_size_t lasIndex)
+      : reader_(reader), index_(index), endIndex_(lasIndex) {}
 
   // Given an element, return an iterator to the first not-null element starting
   // from the element itself.
-  static Iterator initialize(const T& element, vector_size_t endIndex) {
-    auto it = Iterator{element, endIndex};
-
-    // The containier is empty.
-    if (element.index() >= endIndex) {
+  static Iterator initialize(
+      const T* reader_,
+      vector_size_t startIndex,
+      vector_size_t endIndex) {
+    auto it = Iterator{reader_, startIndex, endIndex};
+
+    // The container is empty.
+    if (startIndex >= endIndex) {
       return it;
     }
 
-    if (element.has_value()) {
-      it.currentValue_ = element.value();
+    if (reader_->isSet(startIndex)) {
       return it;
     }
 
+    // Move to next not null.
     it++;
     return it;
   }
 
-  const value_type& operator*() const {
-    return currentValue_;
+  reference operator*() const {
+    // Always return a copy, its guaranteed to be cheap object.
+    return reader_->operator[](index_);
   }
 
-  const value_type* operator->() const {
-    return &currentValue_;
+  pointer operator->() const {
+    // This will work for ViewTypes because they properly overload operator-> to
+    // return this. c++ will keep applying -> recursively on the returned object
+    // until a pointer is returned. This can never be used with primitives.
+    return reader_->operator[](index_);
   }
 
   bool operator<(const Iterator& rhs) const {
-    return this->element_.index() < rhs.element_.index();
+    return index_ < endIndex_;
```

Comment:
Not sure I understand this ; why should index be < endIndex when you are comparing iterators?

Reply:
ops that a bug